src/ffi/ios.rs

```diff
@@ -2,41 +2,110 @@ use core_foundation::{
 
 use crate::{client::Client, DirectoryCache};
+use std::{fs, io};
 
-fn setup_logger() {
-    let subscriber = tracing_fmt::FmtSubscriber::builder()
-        .with_max_level(tracing::Level::DEBUG)
-        .with_writer(std::io::stderr)
-        .finish();
-    tracing::subscriber::set_global_default(subscriber).expect("to be the only logger");
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ArtiRequest {
+    method: String,
+    url: String,
+    headers: HashMap<String, Vec<String>>,
+    body: Vec<u8>,
+    dict_dir: String,
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn call_tls_get(domain_cc: *const c_char) -> CFStringRef {
-    let domain = cstring_to_str(&domain_cc);
+pub unsafe extern "C" fn call_arti(
+    request_json: *const c_char) -> CFStringRef {
+    setup_logger();
+
+    let ret = match _call_arti(cstring_to_str(&request_json)) {
+        Ok(ret) => ret,
+        Err(e) => ReturnStruct::err(e.to_string()),
+    };
+    return to_cf_str(ret.to_json().to_string());
+}
 
-    let client = Client::new(DirectoryCache {
-        tmp_dir: None,
+pub fn _call_arti(request_json: &str) -> Result<ReturnStruct> {
+    let request: ArtiRequest = serde_json::from_str(request_json)?;
+    info!("Request is: {:?}", request);
+
+    let host = Url::parse(&request.url)?.host_str().unwrap().to_string();
+    let mut req = match request.method.as_str() {
+        "GET" => Request::get(request.url),
+        "POST" => Request::post(request.url),
+        "PUT" => Request::put(request.url),
+        "DELETE" => Request::delete(request.url),
+        // TODO: add other verbs of REST
+        // TODO: correctly return error here
+        _ => Request::get(request.url),
+    }
+        .header("Host", host)
+        .version(http::Version::HTTP_10)
+        .body(vec![])?;
+
+    let hm = req.headers_mut();
+    for (key, values) in request.headers {
+        for value in values {
+            hm.append(&HeaderName::from_bytes(&key.as_bytes())?,
+                      HeaderValue::from_bytes(value.as_bytes())?);
+        }
+    }
+
+    info!("Request is: {:?}", req);
+
+    let resp = Client::new(DirectoryCache {
+        tmp_dir: Some(request.dict_dir),
         nodes: None,
         relays: None,
-    });
+    })
+        .send(req)?;
 
-    let req = Request::get(format!("https://{}", domain))
-        .version(http::Version::HTTP_10)
-        .header("Host", domain)
-        .body(vec![])
-        .expect("construct request");
-
-    match client.send(req) {
-        Ok(s) => to_cf_str(format!(
-            "Result is: {:?}",
-            s.map(|raw| String::from_utf8(raw).expect("decode response as utf8"))
-        )),
-        Err(e) => to_cf_str(format!("Error while getting result: {}", e)),
+    ReturnStruct::new(resp.status(), resp.headers(), resp.body())
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ReturnStruct {
+    error: String,
+    status: u16,
+    headers: HashMap<String, Vec<String>>,
```

Comment:
Why is it using a `Vec<String>` as values? From what I remember, it was quite a corner-case (and undefinied in the HTTP spec) to have multiple values for a single header-nam.

Reply:
It is not a corner case at all :) Just make a request, where there are cookies being set :) It is also defined in the HTTP-Spec.

>  Multiple message-header fields with the same field-name MAY be
   present in a message if and only if the entire field-value for that
   header field is defined as a comma-separated list [i.e., #(values)].

[RFC-2616](https://datatracker.ietf.org/doc/html/rfc2616#page-31])

Reply:
Oupsi, I inversed the key and the values :sweat_smile: thanks for catching it.